### ISSUE TITLE
Regression: false negative lint dead_code: dead code is not detected for struct if it implements FromStr

### ISSUE THREAD

**Issue by safinaskar** [rust-lang/rust#142541]
### Code

I tried this code:

```rust
#![deny(dead_code)]

impl std::str::FromStr for Foo {
    type Err = ();
    fn from_str(_s: &str) -> Result<Self, Self::Err> {
        panic!();
    }
}

struct Foo;

fn main() {}
```

I expected to see this happen: compiler should report that struct `Foo` is never constructed

Instead, this happened: compiler didn't report this

### Version it worked on
It most recently worked on: nightly-2019-08-27

### Version with regression
nightly-2019-08-28

### Notes
`cargo-bisect-rustc` reports that the following commits happened between these two versions:

get_commits_between returning commits, len: 5
  commit[0] 2019-08-26: Auto merge of #63901 - estebank:unknown-receiver-type, r=zackmdavis
  commit[1] 2019-08-27: Auto merge of #63926 - Centril:rollup-6kckn9n, r=Centril
  commit[2] 2019-08-27: Auto merge of #63940 - Centril:rollup-47qe9gn, r=Centril
  commit[3] 2019-08-27: Auto merge of #63639 - Mark-Simulacrum:rustdoc-clean-3, r=GuillaumeGomez
  commit[4] 2019-08-27: Auto merge of #63922 - RalfJung:miri, r=nikomatsakis

@rustbot modify labels: +regression-from-stable-to-stable -regression-untriaged +A-lints +L-dead_code +L-false-negative

**Comment by fmease** [comment]
https://github.com/rust-lang/rust/pull/63317

**Comment by safinaskar** [comment]
I still believe current behavior is bug. Struct `Foo` is clearly dead code in my example

**Comment by workingjubilee** [comment]
This only happens in-practice with fieldless types, it seems? This is rejected:
```diff
-struct Foo;
+struct Foo(i32);
```

As is this:
```diff
-struct Foo;
+struct Foo {
+    a: i32,
+}
```

Mind, *this* passes, of course:
```diff
-struct Foo;
+struct Foo {
+    _a: i32,
+}
```

**Comment by safinaskar** [comment]
@workingjubilee , unfortunately, bug is still reproducible with `a: i32`. You just need to add yet another `impl`:

```rust
impl std::str::FromStr for Foo {
    type Err = ();
    fn from_str(_s: &str) -> Result<Self, Self::Err> {
        panic!();
    }
}

impl std::fmt::Debug for Foo {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
        write!(f, "{}", self.a)
    }
}

struct Foo {
    a: i32,
}

fn main() {}
```

(Reproduced on current nightly)

**Comment by mu001999** [comment]
this has been fixed by #142485 ([play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=1efdb86118cc50362a6aa1ebade2649a))

**Comment by mu001999** [comment]
I don't know the exact start time, but for a while, all impl items of trait were always marked as live. After #118257, an impl item is checked based on whether the corresponding trait item is used, but this excludes public trait methods that do not have a receiver (such impl items are always marked as live, which is the case in the current issue). #142485 lifted this restriction.
